src/Workspaces/Core/MSBuild/MSBuild/Options/CodingConventionsServices.cs

```diff
@@ -0,0 +1,60 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Composition.Hosting;
+using System.Linq;
+using System.Reflection;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.MSBuild;
+using Microsoft.VisualStudio.CodingConventions;
+
+namespace Microsoft.CodeAnalysis.MSBuild.Options
+{
+    internal class CodingConventionsServices
+    {
+        private static readonly object s_hostServicesGuard = new object();
+        private static readonly object s_compositionHostGuard = new object();
+
+        private static HostServices s_hostServices;
+        private static CompositionHost s_compositionHost;
+
+        public static readonly ImmutableArray<Assembly> RemoteHostAssemblies =
+            MSBuildMefHostServices.DefaultAssemblies
+                .Add(typeof(MSBuildWorkspace).Assembly)
+                .Add(typeof(ICodingConventionsManager).Assembly);
+
+        public static HostServices HostServices
+        {
+            get
+            {
+                if (s_hostServices != null)
+                {
+                    return s_hostServices;
+                }
+
+                lock (s_hostServicesGuard)
+                {
+                    return s_hostServices ?? (s_hostServices = MefHostServices.Create(CompositionHost));
+                }
+            }
+        }
+
+        internal static CompositionHost CompositionHost
+        {
+            get
+            {
+                if (s_compositionHost != null)
+                {
+                    return s_compositionHost;
+                }
+
+                lock (s_compositionHostGuard)
+                {
+                    var compositionConfiguration = new ContainerConfiguration().WithAssemblies(RemoteHostAssemblies.Distinct());
```

Comment:
Is this safe to put inside a lock? It's unclear if any assembly loading is happening under the hood here. If so probably best to move outside the lock.